src/components/TodoList/TodoList.js

```diff
@@ -1,41 +1,69 @@
 import React from 'react';
 import './TodoList.css';
 
-function handleClick() {
-  console.log(this)
-}
-
 class TodoList extends React.Component {
 
-  // state = {
-  //   list: [
-  //     { id: 1, name: 'Fumer des cannes', active: true },
-  //     { id: 2, name: 'Coder cette shit', active: true },
-  //     { id: 3, name: 'Nager avec des baleines', active: true },
-  //     { id: 4, name: 'Rouler des cannes pour demain', active: true }
-  //   ],
-  //   newTaskName: "",
-  // // }
-  //
-  // handleClick = () => {
-  //   console.log(this)
-  // }
+  state = {
+    list: [ ],
+    newTaskName: "",
+    editTaskName: ""
+  }
 
   deleteTodo(itemToDeleteId) {
     const newList = this.state.list.filter((item) => item.id !== itemToDeleteId)
     this.setState({ list: newList })
   }
 
   createNewTask() {
+    if (this.state.list.length === 0) {
+      this.setState({
+        list: [{ id: 1, name: this.state.newTaskName, active: true, editing: false }],
+        newTaskName: ""
+      })
+      return;
+    }
+
     this.setState({
       list: [
         ...this.state.list,
-        { id: this.state.list[this.state.list.length - 1].id + 1, name: this.state.newTaskName, active: true }
+        { id: this.state.list[this.state.list.length - 1].id + 1, name: this.state.newTaskName, active: true, editing: false}
       ],
-      newTaskName: ""
+      newTaskName: "",
+      editTaskName: ""
     })
   }
 
+  displayEdit(taskId) {
+    const { list } = this.state;
+    const taskIndex = list.findIndex(task => task.id == taskId)
+    const task = list[taskIndex]
+
+    this.setState({ list: [
+      ...list.slice(0, taskIndex),
+    {
+      id: task.id,
+      name: task.name,
+      active: task.active,
+      editing: !task.editing
+    },
+    ...list.slice(taskIndex + 1, list.length)
+    ]})
+  }
+
+  editTask(taskId) {
+    const { list } = this.state;
+    const taskIndex = list.findIndex(task => task.id == taskId)
+    const task = list[taskIndex]
```

Comment:
Considering how we repeat these lines of code, it would make sense to extract them inside a dedicated method function. What about `updateTaskInList` ? It would work like so:
```javascript
updateTaskInList(taskId, taskUpdateObject) {
  const { list } = this.state;
  const taskIndex = list.findIndex(task => task.id == taskId)
  const task = list[taskIndex]

 return [
      ...list.slice(0, taskIndex),
    { ...task, ...taskUpdateObject },
    ...list.slice(taskIndex + 1, list.length)
  ]
}

// Would be used like so
editTask(taskId) {
  this.setState({
    list: updateTaskInList(taskId, { editing: true }),
    editTaskName: ""
  })
}
```

Reply:
So, as you can see, `updateTaskInList` does not directly update the state of the component, instead, it generates a new list which has the update that you want to do. 
The method that calls it, in this example `editTask`, is responsible to update the state. That way, we keep control on what is really going on inside `editTask`. `updateTaskInList ` does not have any "side effects", that means, impact outside of its scope. It takes arguments, return a new list, but did not modify aside from this.